### ISSUE TITLE
Cache cluster IPs and use if original RethinkDB host is unreachable

### ISSUE THREAD

**Issue by colegleason** [rethinkdb/rethinkdb-go#105]
I have a case where I have connected to a machine in a cluster of 3 rethinkdb instances. When that machines goes down, all the tables are still available (due to replicas, etc). It would be awesome if gorethink could cache the IPs of the other nodes in the cluster and failover if needed.

cc @philips

**Comment by dancannon** [comment]
I agree this is definitely something that can be included in the driver and something I have had on my todo list. 

I will try to get at least some kind of basic cluster support ready for the stable RethinkDB release. Once I have looked into the kind of features that need to be included I will reply to this issue.

**Comment by dancannon** [comment]
After researching how I would do this I think that it makes sense to wait for RethinkDB 1.14(rethinkdb/rethinkdb#2663) as there are a number of useful features that have been proposed that would makes this feature much more useful.

**Comment by timmaxw** [comment]
The features proposed in https://github.com/rethinkdb/rethinkdb/issues/2663 probably won't ship for another month or two. They will not be included in RethinkDB 1.14.

**Comment by nkev** [comment]
Is there any development on the auto-failover front?

**Comment by dancannon** [comment]
Yes, I have started the auto-failure + cluster support code last week (I used the go-hostpool lib to easily add the auto-failover), I hope to be adding host discovery soon using the cluster management API adding in RethinkDB 1.16.

**Comment by dancannon** [comment]
I have got an initial version working, it is not yet complete but I would appreciate any feedback if anybody has some time to spend reviewing it.

Currently all nodes in your cluster need to be started with the `--canonical-address` manually specified. This is because by default the `server_status` table returns both internal and external canonical addresses.

To connect to a cluster: 

``` go
cluster, err := ConnectClusterWithOpts(ConnectOpts{
    DiscoverHosts: true,
}, "localhost:28015", "localhost:28016")
```

If you want the driver to look for other nodes in the cluster you can use:

``` go
cluster, err := ConnectCluster("localhost:28015", "localhost:28016")
```

I have also simplified the existing connection code:

``` go
cluster, err := Connect("localhost:28015")
```

**Comment by dancannon** [comment]
@colegleason are you still using the driver/interested in this functionality? If so could you have a look at the changes in https://github.com/dancannon/gorethink/pull/167

**Comment by nkev** [comment]
@dancannon I think everyone would be interested in auto-failover built right into the driver! :+1: 
I'm right in the middle of a project, otherwise I would test it too.

**Comment by colegleason** [comment]
Sorry @dancannon, but I am not using RethinkDB anymore!

**Comment by dancannon** [comment]
@colegleason ok, thanks for the suggestion anyway.

**Comment by dancannon** [comment]
This can now be closed at the latest version (v0.7.0) has support for connecting to multiple nodes + host discovery.

Currently queries are distributed between the nodes using round robin, I plan to improve this in the near future.
